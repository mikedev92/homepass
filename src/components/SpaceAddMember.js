// @flow
import React from 'react';
import {
  List,
  Thumbnail,
  ListItem,
  Container,
  Header,
  Left,
  Title,
  Body,
  Content,
  Right,
  Icon,
  Button,
  Text,
} from 'native-base';
import { compose, withState, withHandlers } from 'recompose';
import Search from 'react-native-search-box';
import { graphql } from 'react-apollo';
import { lowerCase } from 'lodash';
import gql from 'graphql-tag';

type Props = {
  data: ApolloData<Space>,
  navigation: Navigation,
  toggleMember: string => void,
  setSearchText: string => void,
  selectedMembers: {
    [string]: boolean,
  },
  searchText: string,
  addMembersToSpace: ({ [string]: boolean }) => void,
};

function SpaceAddMember({
  data,
  navigation,
  searchText,
  setSearchText,
  toggleMember,
  selectedMembers,
  addMembersToSpace,
}: Props) {
  if (!data.node || !data.node.suggestedMembers) {
    return null;
  }
  const suggestedMembers = data.node.suggestedMembers.filter(
    user => user.name && lowerCase(user.name).includes(lowerCase(searchText)),
  );
  return (
    <Container>
      <Header bordered>
        <Left>
          <Button onPress={() => navigation.goBack()} transparent primary>
            <Icon name="arrow-back" />
          </Button>
        </Left>
        <Body><Title>Add Members</Title></Body>
        <Right>
          <Button onPress={() => addMembersToSpace(selectedMembers)} transparent primary>
            <Text primary>Done</Text>
          </Button>
        </Right>
      </Header>
      <Content keyboardShouldPersistTaps="handled">
        <List>
          <Search
            backgroundColor="transparent"
            onChangeText={setSearchText}
            titleCancelColor="black"
          />
          {suggestedMembers.map(
            (user, index) =>
              user.userId &&
              <ListItem
                onPress={() => user.userId && toggleMember(user.userId)}
                avatar
                icon
                key={user.userId}
                last={suggestedMembers.length - 1 === index}
              >
                {user.imageUrl &&
                  <Left>
                    <Thumbnail small source={{ uri: user.imageUrl }} />
                  </Left>}
                <Body>
                  <Text>
                    {user.name}
                  </Text>
                </Body>
                <Right>
                  <Icon primary active={selectedMembers[user.userId]} name="checkmark-circle" />
                </Right>
              </ListItem>,
          )}
        </List>
      </Content>
    </Container>
  );
}

const SpaceAddMemberQuery = gql`
  query SpaceAddMemberQuery($id: ID!) {
    node(id: $id) {
      id
      ... on Space {
        spaceId
        suggestedMembers {
          id
          userId
          name
          imageUrl
        }
      }
    }
  }
`;

const AddMemberToSpaceMutation = gql`
  mutation($input: AddSpaceMemberInput!) {
    addSpaceMember(input: $input) {
      space {
        id
        members {
          edges {
            node {
              id
              user {
                id
              }
            }
          }
        }
        suggestedMembers {
          id
          userId
          name
          imageUrl
        }
      }
    }
  }
`;

export default compose(
  graphql(SpaceAddMemberQuery, {
    options: props => ({ variables: { id: props.navigation.state.params.id } }),
  }),
  graphql(AddMemberToSpaceMutation, {
    props: ({ mutate, ownProps }) => ({
      addMembersToSpace: async (userIds) => {
        const mutations = Object.keys(userIds).map(
          userId =>
            userIds[userId] &&
            mutate({
              variables: {
                input: {
                  userId,
                  spaceId: ownProps.data.node.spaceId,
                },
              },
            }),
        );
        await Promise.all(mutations);
        return ownProps.navigation.goBack();
      },
    }),
  }),
  withState('searchText', 'setSearchText', ''),
  withState('selectedMembers', 'setSelectedMembers', {}),
  withHandlers({
    toggleMember: ({ setSelectedMembers, selectedMembers }) => id =>
      setSelectedMembers({
        ...selectedMembers,
        [id]: !selectedMembers[id],
      }),
  }),
)(SpaceAddMember);

export {
  SpaceAddMember as Raw,
};
