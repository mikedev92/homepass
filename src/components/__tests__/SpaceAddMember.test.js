import React from 'react';
import { shallow } from 'enzyme';
import { ListItem } from 'native-base';
import { Raw as SpaceAddMember } from '../SpaceAddMember';

const testProps = [
  {
    data: {
      loading: false,
      node: {
        id: '123',
        spaceId: '123',
        suggestedMembers: [
          {
            id: 'abc',
            userId: 'adc',
            name: 'my image',
          },
          {
            id: 'cde',
            userId: 'cde',
            name: 'my image 2',
          },
          {
            id: '345',
            userId: '345',
            name: 'my image 3',
          },
        ],
      },
    },
    selectedMembers: { abc: true, cde: true }, // eslint-disable-line
  },
];
describe('<SpaceAddMember />', () => {
  it('matches snapshots', () => {
    testProps.forEach((props) => {
      const baseProps = {
        toggleMember: jest.fn(),
      };
      const wrapper = shallow(<SpaceAddMember {...baseProps} {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('calls toggleMember correctly', () => {
    testProps.forEach((props) => {
      const baseProps = {
        toggleMember: jest.fn(() => jest.fn()),
      };
      const wrapper = shallow(<SpaceAddMember {...baseProps} {...props} />);
      const listItems = wrapper.find(ListItem);
      listItems.forEach(listItem => listItem.prop('onPress')());
      expect(baseProps.toggleMember).toHaveBeenCalledTimes(listItems.length);
    });
  });
});
