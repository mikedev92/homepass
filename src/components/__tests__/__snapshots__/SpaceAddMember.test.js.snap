// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpaceAddMember /> matches snapshots 1`] = `
<Styled(Container)>
  <Styled(Header)
    bordered={true}
  >
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="arrow-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body)>
      <Styled(Title)>
        Add Members
      </Styled(Title)>
    </Styled(Body)>
    <Styled(Right)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Text)
          primary={true}
        >
          Done
        </Styled(Text)>
      </Styled(Button)>
    </Styled(Right)>
  </Styled(Header)>
  <Styled(Content)
    keyboardShouldPersistTaps="handled"
  >
    <Styled(List)>
      <Search
        backgroundColor="transparent"
        blurOnSubmit={false}
        editable={true}
        keyboardShouldPersist={false}
        placeholderCollapsedMargin={15}
        placeholderExpandedMargin={20}
        searchIconCollapsedMargin={25}
        searchIconExpandedMargin={10}
        shadowColor="#000"
        shadowOffsetHeightCollapsed={2}
        shadowOffsetHeightExpanded={4}
        shadowOffsetWidth={0}
        shadowOpacityCollapsed={0.12}
        shadowOpacityExpanded={0.24}
        shadowRadius={4}
        shadowVisible={false}
        titleCancelColor="black"
      />
      <Styled(ListItem)
        avatar={true}
        icon={true}
        last={false}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            my image
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Icon)
            name="checkmark-circle"
            primary={true}
          />
        </Styled(Right)>
      </Styled(ListItem)>
      <Styled(ListItem)
        avatar={true}
        icon={true}
        last={false}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            my image 2
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Icon)
            active={true}
            name="checkmark-circle"
            primary={true}
          />
        </Styled(Right)>
      </Styled(ListItem)>
      <Styled(ListItem)
        avatar={true}
        icon={true}
        last={true}
        onPress={[Function]}
      >
        <Styled(Body)>
          <Styled(Text)>
            my image 3
          </Styled(Text)>
        </Styled(Body)>
        <Styled(Right)>
          <Styled(Icon)
            name="checkmark-circle"
            primary={true}
          />
        </Styled(Right)>
      </Styled(ListItem)>
    </Styled(List)>
  </Styled(Content)>
</Styled(Container)>
`;
